8319137: release _object in ObjectMonitor dtor to avoid races

pchilano opened the pull request:
Please review the following fix which moves the release of the _object WeakHandle back to the ObjectMonitor dtor. Releasing it earlier as we do today introduces a race which can lead to crashes. Tested running tiers123 in mach5.

This fix is an effective backout of [JDK-8256302](https://bugs.openjdk.org/browse/JDK-8256302).

Thanks,
Patricio

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8319137](https://bugs.openjdk.org/browse/JDK-8319137): release _object in ObjectMonitor dtor to avoid races (**Bug** - P2)


### Reviewers
 * [Stefan Karlsson](https://openjdk.org/census#stefank) (@stefank - **Reviewer**)
 * [Axel Boldt-Christmas](https://openjdk.org/census#aboldtch) (@xmas92 - **Reviewer**) ⚠️ Review applies to [3cfea97d](https://git.openjdk.org/jdk/pull/16738/files/3cfea97d74d19c7e06bc2e801fcd0b896b5d0186)
 * [Aleksey Shipilev](https://openjdk.org/census#shade) (@shipilev - **Reviewer**) ⚠️ Review applies to [3cfea97d](https://git.openjdk.org/jdk/pull/16738/files/3cfea97d74d19c7e06bc2e801fcd0b896b5d0186)
 * [Daniel D. Daugherty](https://openjdk.org/census#dcubed) (@dcubed-ojdk - **Reviewer**)
 * [Erik Österlund](https://openjdk.org/census#eosterlund) (@fisk - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16738/head:pull/16738` \
`$ git checkout pull/16738`

Update a local copy of the PR: \
`$ git checkout pull/16738` \
`$ git pull https://git.openjdk.org/jdk.git pull/16738/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16738`

View PR using the GUI difftool: \
`$ git pr show -t 16738`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16738.diff">https://git.openjdk.org/jdk/pull/16738.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16738#issuecomment-1819247076)

stefank reviewed (approved):
Looks good.

stefank commented on src/hotspot/share/runtime/objectMonitor.cpp:
```diff
@@ -276,10 +276,7 @@ ObjectMonitor::ObjectMonitor(oop object) :
-    // Release object's oop storage if it hasn't already been done.
-    release_object();
-  }
+  release_object();
```
I think it would be nice to move the implementation of release_object over to the .cpp file. Alternatively, inline the implementation straight into the destructor. If you don't want to do it in this PR I can do it as a separate change.

pchilano replied:
Done.

shipilev commented:
How much of this is a revert of https://github.com/openjdk/jdk/commit/6402004852ccc993bb6e879c49fb71e2afdcdfbe? Should we revert back to that state?

shipilev commented:
It would be good to also name the issue by the actual thing that we are doing, rather than the symptom we are seeing in one of the tests.

xmas92 reviewed (approved):
Lgtm.

Could potentially remove the `if (_object.is_null())` checks in `ObjectMonitor::object` and `ObjectMonitor::object_peek`. Should never take those branches.

pchilano commented:
> It would be good to also name the issue by the actual thing that we are doing, rather than the symptom we are seeing in one of the tests.
>
I changed the bug title.

shipilev commented:
> > How much of this is a revert of [6402004](https://github.com/openjdk/jdk/commit/6402004852ccc993bb6e879c49fb71e2afdcdfbe)? Should we revert back to that state?
> 
> It's pretty much a revert of that. Now only the deflater thread can call this code so I removed the extra branch in the delete loop.

Then I believe a cleaner way to do this is to make a clean backout of [JDK-8256302](https://bugs.openjdk.org/browse/JDK-8256302), or am I missing something?

pchilano commented:
> > > How much of this is a revert of [6402004](https://github.com/openjdk/jdk/commit/6402004852ccc993bb6e879c49fb71e2afdcdfbe)? Should we revert back to that state?
> > 
> > 
> > It's pretty much a revert of that. Now only the deflater thread can call this code so I removed the extra branch in the delete loop.
> 
> Then I believe a cleaner way to do this is to make a clean backout of [JDK-8256302](https://bugs.openjdk.org/browse/JDK-8256302), or am I missing something?
>
So with the current changes this fix is almost a backout except for the extra branch I removed in the loop given that now only the deflater thread can call this, and the fact that the delete loop was factored out into a separate static method. Can't this be considered a backout and rename it as such?

shipilev commented:
> So with the current changes this fix is almost a backout except for the extra branch I removed in the loop given that now only the deflater thread can call this, and the fact that the delete loop was factored out into a separate static method. Can't this be considered a backout and rename it as such?

Probably not worth the hassle of doing the backout and resolve the conflicts there. But it should be clearly said somewhere (e.g. both in PR body and the JBS issue) that the issue is with [JDK-8256302](https://bugs.openjdk.org/browse/JDK-8256302) and the fix is the effective backout of [JDK-8256302](https://bugs.openjdk.org/browse/JDK-8256302).

pchilano commented:
> Lgtm.
> 
> Could potentially remove the `if (_object.is_null())` checks in `ObjectMonitor::object` and `ObjectMonitor::object_peek`. Should never take those branches.
>
Right, removed.

dcubed-ojdk commented:
You could do this as "[BACKOUT] JDK-8256302 releasing oopStorage when deflating allows for faster deleting"
and then merge that [BACKOUT] into the main line and resolve all conflicts. There have been quite a few follow
on changes that would have to be resolved.

That would be "cleaner" from a process POV, but I don't think it is necessary.

Please add a note to [JDK-8256302](https://bugs.openjdk.org/browse/JDK-8256302) making it clear
that fix has been effectively backed out by this fix.

pchilano commented:
> You could do this as "[BACKOUT] JDK-8256302 releasing oopStorage when deflating allows for faster deleting" and then merge that [BACKOUT] into the main line and resolve all conflicts. There have been quite a few follow on changes that would have to be resolved.
> 
> That would be "cleaner" from a process POV, but I don't think it is necessary.
> 
> Please add a note to [JDK-8256302](https://bugs.openjdk.org/browse/JDK-8256302) making it clear that fix has been effectively backed out by this fix.
>
Sounds good, I'll do that after integrating this.

dcubed-ojdk reviewed (approved):
This is the equivalent of a [BACKOUT] of [JDK-8256302](https://bugs.openjdk.org/browse/JDK-8256302)
with a couple of adaptations in keeping with the current state of the code:
- the static `delete_monitors()` function is kept
- only the MonitorDeflationThread deflates ObjectMonitors any more so some logic that predates JDK-8256302 is no longer needed

I recommend redoing the Tier[1-3] round of testing. I don't
think you need to do Tier[4-6] testing with this [BACKOUT].

JDK-8256302 was integrated in jdk-21+26 (2023.06.07).
JDK-8319137 was found in jdk-22+21 stress testing (2023.10.28)
almost 5 months later so shaking out bugs in this area of code is
difficult.

fisk reviewed (approved):
Looks good.

dholmes-ora commented:
> Lgtm.
> 
> Could potentially remove the `if (_object.is_null())` checks in `ObjectMonitor::object` and `ObjectMonitor::object_peek`. Should never take those branches.

@xmas92 is your comment/suggestion now invalidated by the issue we discovered with JNI monitor locking and rclaimed objects?

xmas92 commented:
> > Lgtm.
> > Could potentially remove the `if (_object.is_null())` checks in `ObjectMonitor::object` and `ObjectMonitor::object_peek`. Should never take those branches.
> 
> @xmas92 is your comment/suggestion now invalidated by the issue we discovered with JNI monitor locking and rclaimed objects?

`WeakHandle::is_null` refers to the pointer into oop storage. So the reference, not the referent.

xmas92 commented:
> > > Lgtm.
> > > Could potentially remove the `if (_object.is_null())` checks in `ObjectMonitor::object` and `ObjectMonitor::object_peek`. Should never take those branches.
> > 
> > 
> > @xmas92 is your comment/suggestion now invalidated by the issue we discovered with JNI monitor locking and rclaimed objects?
> 
> `WeakHandle::is_null` refers to the pointer into oop storage.

It is interesting that we have the `is_empty` method as well that does the same thing. Regardless these names should probably be cleared up. It is unclear if it refers to the `*oop` or the `oop`.

dcubed-ojdk commented:
@robehn - [JDK-8256302](https://bugs.openjdk.org/browse/JDK-8256302) was originally
your idea that came up from another code review. Did you want to chime in on this review?

robehn commented:
> @robehn - [JDK-8256302](https://bugs.openjdk.org/browse/JDK-8256302) was originally your idea that came up from another code review. Did you want to chime in on this review?

If there in practice was no performance benefit reverting seems like good option.
As it's been reviewed by several heavy hitters I'm all good!

pchilano commented:
@dcubed-ojdk I re-run Tiers[1-3] and all tests passed except an unrelated DNS test failure.

pchilano commented:
Alright, thanks everyone for the reviews!

pchilano commented:
/integrate
